File: src/etch/bindings/csharp/support/StubBase.cs
///$Id$
///
///Created by Champakesan, Badri Narayanan on Jun 25, 2007.
///
///Copyright (c) 2007 Cisco Systems, Inc. All rights reserved. 


using System;
using Etch.Msg;
using Etch.Transport;

namespace Etch.Support{

    public class StubBase : Session, MessageHandler
    {   
	    /// <summary>Constructs the StubBase.</summary>
	    /// <param name="obj">obj</param>
        /// <param name="queued">thread pool used to run AsyncReceiverMode.QUEUED methods.</param>
        /// <param name="free">thread pool used to run AsyncReceiverMode.FREE methods.</param>
	    public StubBase( MessageSource src, Object obj, Pool queued, Pool free )
	    {
            _src = src;
		    _obj = obj;
            _queued = queued;
            _free = free;
	    }
    	
	    private readonly Object _obj;

        /// <summary>
        /// The thread pool used to run AsyncReceiverMode.QUEUED methods.
        /// </summary>
        protected readonly Pool _queued;

        /// <summary>
        /// The thread pool used to run AsyncReceiverMode.FREE methods.
        /// </summary>
        protected readonly Pool _free;

        protected readonly MessageSource _src;

        public virtual bool Message( Who sender, Message msg )
        {
            return false;
        }

        public Object SessionQuery( Object query )
        {
            if ( _obj is ObjSession )
                return ( ( ObjSession ) _obj )._SessionQuery( query );
            throw new Exception( "unknown query "+query );
        }

        public void SessionControl( Object control, Object value )
	    {
            if ( _obj is ObjSession )
                ( ( ObjSession ) _obj )._SessionControl( control, value );
            else
		        throw new Exception( "unknown control "+control );
	    }

	    public void SessionNotify( Object eventObj )
	    {
            if ( _obj is ObjSession )
                ( ( ObjSession ) _obj )._SessionNotify( eventObj );
	    }
    }
}

File: src/etch/bindings/java/support/StubBase.java
/* $Id$
 *
 * Created by sccomer on Mar 29, 2007.
 *
 * Copyright (c) 2007 Cisco Systems, Inc. All rights reserved.
 */

package etch.bindings.java.support;

import metreos.core.Who;
import etch.bindings.java.msg.Message;
import etch.bindings.java.transport.Session;

/**
 * Base class of stub implementations.
 */
public class StubBase implements MessageHandler, Session
{
	/**
	 * Constructs the StubBase.
	 * @param src the message source.
	 * @param obj the target of decoded messages.
	 * @param queued thread pool used to run AsyncReceiverMode.QUEUED methods.
	 * @param free thread pool used to run AsyncReceiverMode.FREE methods.
	 */
	public StubBase( MessageSource src, Object obj, Pool queued, Pool free )
	{
		this._src = src;
		this._obj = obj;
		this._queued = queued;
		this._free = free;
	}
	
	/**
	 * The message source.
	 */
	protected final MessageSource _src;
	
	@SuppressWarnings("unused")
	private final Object _obj;
	
	/**
	 * The thread pool used to run AsyncReceiverMode.QUEUED methods.
	 */
	protected final Pool _queued;
	
	/**
	 * The thread pool used to run AsyncReceiverMode.FREE methods.
	 */
	protected final Pool _free;

	public boolean message( Who sender, Message msg ) throws Exception
	{
		return false;
	}

	public Object sessionQuery( Object query )
	{
		// TODO reflect sessionQuery to service implementation object.
		throw new UnsupportedOperationException( "unknown query "+query );
	}

	public void sessionControl( Object control, Object value )
	{
		// TODO reflect sessionControl to service implementation object.
		throw new UnsupportedOperationException( "unknown control "+control );
	}

	public void sessionNotify( Object event )
	{
		// TODO reflect sessionNotify to service implementation object.
	}
}
